client/src/main/java/io/cloudsoft/winrm4j/client/WinRmClient.java

```diff
@@ -244,43 +258,79 @@ private int receiveCommand(String commandId, Writer out, Writer err) {
-            ReceiveResponse receiveResponse = winrm.receive(receive, RESOURCE_URI, MAX_ENVELOPER_SIZE, operationTimeout, locale, shellSelector);
-            List<StreamType> streams = receiveResponse.getStream();
-            for (StreamType s : streams) {
-                byte[] value = s.getValue();
-                if (value == null) continue;
-                if (out != null && "stdout".equals(s.getName())) {
-                    try {
-                        //TODO use passed locale?
-                        if (value.length > 0) {
-                            out.write(new String(value));
-                        }
-                        if (Boolean.TRUE.equals(s.isEnd())) {
-                            out.close();
-                        }
-                    } catch (IOException e) {
-                        throw new IllegalStateException(e);
+
+            try {
+                numberOfReceiveCalls++;
+                ReceiveResponse receiveResponse = winrm.receive(receive, RESOURCE_URI, MAX_ENVELOPER_SIZE, operationTimeout, locale, shellSelector);
+                getStreams(receiveResponse, out, err);
+
+                CommandStateType state = receiveResponse.getCommandState();
+                if (COMMAND_STATE_DONE.equals(state.getState())) {
+                    return state.getExitCode().intValue();
+                } else {
+                    LOG.debug("{} is not done. Response it received: {}", this, receiveResponse);
+                }
+            } catch (SOAPFaultException soapFault) {
+                /**
+                 * If such Exception which has a code 2150858793 the client is expected to again trigger immediately a receive request.
+                 * https://msdn.microsoft.com/en-us/library/cc251676.aspx
+                 */
+
+                try {
+                    if (soapFault.getFault().getDetail().getFirstChild().getAttributes().getNamedItem("Code").getNodeValue().equals(WSMAN_FAULT_CODE_OPERATION_TIMEOUT_EXPIRED)) {
+                        LOG.trace("winrm client {} received an operation response {}", this, soapFault);
+                    } else {
+                        throw soapFault;
                     }
+                } catch (NullPointerException e) {
+                    throw soapFault;
                 }
-                if (err != null && "stderr".equals(s.getName())) {
-                    try {
-                        //TODO use passed locale?
-                        if (value.length > 0) {
-                            err.write(new String(value));
-                        }
-                        if (Boolean.TRUE.equals(s.isEnd())) {
-                            err.close();
-                        }
-                    } catch (IOException e) {
-                        throw new IllegalStateException(e);
+            }
+        }
+    }
+
+    public int getNumberOfReceiveCalls() {
```

Comment:
It's weird to have internal state exposed just for the tests. If you'd like to verify that re-connect is working could use a local server (http components has a nice testing server for example) to verify.
A local server would be nice in general for doing non-live tests with pre-defined responses.